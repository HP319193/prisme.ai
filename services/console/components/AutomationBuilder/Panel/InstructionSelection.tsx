import { useTranslation } from "next-i18next";
import Image from "next/image";
import { Button } from "primereact/button"
import { InputText } from "primereact/inputtext"
import { FC, useMemo, useState } from "react"
import { useAutomationBuilder } from "../context";

export interface InstructionSelectionProps {
  onSubmit: (key: string) => void;
}

export const InstructionSelection: FC<InstructionSelectionProps> = ({ onSubmit }) => {
  const { t } = useTranslation('workspaces');
  const { instructionsSchemas } = useAutomationBuilder();
  const [search, setSearch] = useState('');
  const filteredInstructions = useMemo(() => {
    return instructionsSchemas.reduce<[string, string, string[]][]>((prev, [name, list, more]) => {
      const matching = search ? Object.keys(list).filter(a => a.match(search)) : Object.keys(list)
      if (!matching) return prev;
      return [
        ...prev,
        [name, more.icon, matching]
      ];
    }, []);
  }, [instructionsSchemas, search])

  return (
    <>
      <div className="flex flex-1 align-items-stretch flex-column mb-4">
        <div>{t('automations.edit.select')}</div>
        <span className="p-input-icon-left p-input-icon-right">
          <i className="pi pi-search" />
          <InputText
            value={search}
            onChange={(e) => setSearch(e.target.value)}
            placeholder={t('automations.instruction.search')}
            autoFocus
          />
          <i><button onClick={() => setSearch('')} style={{ background: 'none', border: 0, color: 'inherit' }}><i className="pi pi-times" /></button></i>
        </span>
      </div>
      {filteredInstructions.map(([section, icon, instructionsInSection]) => (
        <div key={section} className="flex flex-1 flex-column mb-4">
          <div className="flex align-items-center">
            <div className="mr-2">
              <Image src={icon} width={16} height={16} alt={section} />
            </div>
            {section}
          </div>
          {instructionsInSection.map(key => (
            <Button
              key={key}
              onClick={() => onSubmit(key)}
              className="p-button-outlined my-1"
            >{key}</Button>
          ))}
        </div>
      ))}
    </>
  )
}

export default InstructionSelection
