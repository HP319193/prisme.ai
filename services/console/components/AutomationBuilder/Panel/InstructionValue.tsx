import { FC, useCallback, useMemo, useState } from 'react';
import Fieldset from '../../../layouts/Fieldset';
import { useTranslation } from 'next-i18next';
import {
  FieldProps,
  getSchemaFormLabel,
  SchemaForm,
  SchemaFormDescription,
} from '@prisme.ai/design-system';
import { CodeEditorInline } from '../../CodeEditor/lazy';
import { useField } from 'react-final-form';
import FieldContainerWithRaw from '../../FieldContainerWithRaw';
import useSchema from '../../SchemaForm/useSchema';
import { useWorkspace } from '../../../layouts/WorkspaceLayout';
import usePages from '../../PagesProvider/context';

interface InstructionValueProps {
  instruction: string;
  value: any;
  schema?: any;
  onChange: (values: any) => void;
}

const EmptyButtons: any[] = [];
const FieldAny = ({ schema, name, label }: FieldProps) => {
  const field = useField(name);
  const [value, setValue] = useState(
    typeof field.input.value === 'string'
      ? field.input.value
      : JSON.stringify(field.input.value, null, '  ')
  );
  const onChange = useCallback(
    (value: string) => {
      setValue(value);
      try {
        const json = JSON.parse(value);
        field.input.onChange(json);
      } catch {
        field.input.onChange(value);
      }
    },
    [field.input]
  );

  return (
    <div className="flex flex-1 flex-col my-2">
      <SchemaFormDescription text={schema.description}>
        <label className="text-[10px] text-gray">
          {label || schema.title || getSchemaFormLabel(name)}
        </label>
        <CodeEditorInline
          value={value}
          onChange={onChange}
          mode="json"
          style={{ flex: 'auto' }}
        />
      </SchemaFormDescription>
    </div>
  );
};

const components = {
  FieldAny,
  FieldContainer: FieldContainerWithRaw,
};

export const InstructionValue: FC<InstructionValueProps> = ({
  instruction,
  value,
  schema = {},
  onChange,
}) => {
  const { workspace } = useWorkspace();
  const { pages } = usePages();
  const { t } = useTranslation('workspaces');

  const appInstance = useMemo(() => {
    if (!workspace.imports) return workspace.config;
    const [appName] = instruction.split(/\./);
    if (!workspace.imports[appName]) return workspace.config;
    return workspace.imports[appName].config || {};
  }, [instruction, workspace.config, workspace.imports]);

  const { extractSelectOptions } = useSchema({
    config: appInstance,
    automations: workspace.automations,
    pages: pages.get(workspace.id),
  });

  const cleanedSchema = useMemo(() => {
    const cleaned = {
      ...schema,
      title: t('automations.instruction.description', {
        context: instruction,
        default: schema.description,
      }),
    };
    if (instruction === 'repeat') {
      cleaned['ui:options'] = {
        oneOf: {
          options: [
            {
              label: 'automations.instruction.form.repeat.on.label',
              index: 0,
              value: {
                until: undefined,
              },
            },
            {
              label: 'automations.instruction.form.repeat.until.label',
              index: 1,
              value: {
                on: undefined,
              },
            },
          ],
        },
      };
    }
    return cleaned;
  }, [instruction, schema, t]);

  const locales = useMemo(
    () => ({
      addItem: t('automations.instruction.form.addItem'),
      addProperty: t('automations.instruction.form.addProperty'),
      propertyKey: t('automations.instruction.form.propertyKey'),
      propertyValue: t('automations.instruction.form.propertyValue'),
      removeItem: t('automations.instruction.form.removeItem'),
      removeProperty: t('automations.instruction.form.removeProperty'),
      uploadLabel: t('automations.instruction.form.uploadLabel'),
      uploadRemove: t('automations.instruction.form.uploadRemove'),
    }),
    [t]
  );

  if (!schema) return null;

  return (
    <Fieldset
      legend={t('automations.instruction.label', { context: instruction })}
      hasDivider={false}
    >
      <SchemaForm
        schema={cleanedSchema}
        onChange={onChange}
        initialValues={value}
        buttons={EmptyButtons}
        components={components}
        locales={locales}
        utils={{ extractSelectOptions }}
      />
    </Fieldset>
  );
};

export default InstructionValue;
