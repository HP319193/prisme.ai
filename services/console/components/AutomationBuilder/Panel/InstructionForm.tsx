import { FC, useCallback, useEffect, useState } from 'react';
import InstructionSelection from './InstructionSelection';
import InstructionValue from './InstructionValue';
import { Schema } from '../../SchemaForm/types';
import { useAutomationBuilder } from '../context';
import { ArrowLeftOutlined } from '@ant-design/icons';

const getDefaultValue = (type: string) => {
  switch (type) {
    case 'array':
      return [];
    case 'object':
    default:
      return {};
  }
};
interface InstructionFormProps {
  instruction?: Prismeai.Instruction;
  onSubmit: (i: Prismeai.Instruction) => void;
}

export const InstructionForm: FC<InstructionFormProps> = ({
  instruction,
  onSubmit,
}) => {
  const { getSchema } = useAutomationBuilder();
  const [currentInstruction] = Object.keys(instruction || {});
  const [edit, setEdit] = useState<{
    instruction: string;
    value: any;
    schema: Schema | null;
  }>({
    instruction: currentInstruction,
    value:
      currentInstruction &&
      instruction &&
      instruction[currentInstruction as keyof typeof instruction],
    schema: getSchema(currentInstruction),
  });

  const setInstruction = useCallback(
    (instructionName: string) => {
      const schema = getSchema(instructionName);
      if (!schema.properties) {
        onSubmit({
          [instructionName]: getDefaultValue(schema.type || ''),
        });
        return;
      }
      setEdit({
        instruction: instructionName,
        value: {},
        schema,
      });
    },
    [getSchema, onSubmit]
  );

  useEffect(() => {
    const [currentInstruction] = Object.keys(instruction || {});
    setEdit({
      instruction: currentInstruction,
      value:
        currentInstruction &&
        instruction &&
        instruction[currentInstruction as keyof typeof instruction],
      schema: getSchema(currentInstruction),
    });
  }, [getSchema, instruction, setInstruction]);

  const setInstructionValue = useCallback(
    (values: any) => {
      if (!edit.instruction) return;
      onSubmit({
        [edit.instruction]: values,
      });
    },
    [edit.instruction, onSubmit]
  );

  const unsetInstruction = useCallback(() => {
    setEdit({
      instruction: '',
      value: {},
      schema: null,
    });
  }, []);

  return (
    <div className="flex flex-1 flex-col h-full overflow-x-auto">
      {!instruction && edit.instruction && (
        <button
          onClick={unsetInstruction}
          className="absolute top-[24px] left-[1rem]"
        >
          <ArrowLeftOutlined />
        </button>
      )}
      {!edit ||
        (!edit.instruction && (
          <InstructionSelection onSubmit={setInstruction} />
        ))}
      {edit && edit.instruction && (
        <InstructionValue
          instruction={edit.instruction}
          value={edit.value}
          schema={edit.schema}
          onSubmit={setInstructionValue}
        />
      )}
    </div>
  );
};

export default InstructionForm;
