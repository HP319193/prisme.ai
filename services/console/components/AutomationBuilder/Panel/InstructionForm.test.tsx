import InstructionForm from './InstructionForm';
import renderer, { act } from 'react-test-renderer';
import InstructionSelection from './InstructionSelection';
import InstructionValue from './InstructionValue';

jest.mock('../context', () => {
  const mock = {
    getSchema: (name: string) => {
      if (name === 'wait') return {};
      if (name === 'emit')
        return {
          type: 'object',
        };
      if (name === 'conditions')
        return {
          type: 'object',
        };
      if (name === 'all')
        return {
          type: 'array',
        };
      return {};
    },
    instructionsSchemas: [],
  };
  return {
    useAutomationBuilder: () => mock,
  };
});

it('should render', () => {
  const onSubmit = jest.fn();
  const onChange = jest.fn();
  const root = renderer.create(
    <InstructionForm onSubmit={onSubmit} onChange={onChange} />
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it('should set instruction without params', async () => {
  jest.useFakeTimers();
  const onSubmit = jest.fn();
  const onChange = jest.fn();
  const root = renderer.create(
    <InstructionForm onSubmit={onSubmit} onChange={onChange} />
  );

  await act(async () => {
    await true;
    jest.runAllTimers();
  });

  act(() => {
    root.root.findByType(InstructionSelection).props.onSubmit('wait');
  });

  expect(onSubmit).toHaveBeenCalledWith({
    wait: {},
  });
  onSubmit.mockClear();

  act(() => {
    root.root.findByType(InstructionSelection).props.onSubmit('conditions');
  });

  expect(onSubmit).toHaveBeenCalledWith({
    conditions: {},
  });
  onSubmit.mockClear();

  act(() => {
    root.root.findByType(InstructionSelection).props.onSubmit('all');
  });

  expect(onSubmit).toHaveBeenCalledWith({
    all: [],
  });
});

it('should set instruction with params', async () => {
  jest.useFakeTimers();
  const onChange = jest.fn();
  const onSubmit = jest.fn();
  const root = renderer.create(
    <InstructionForm onChange={onChange} onSubmit={onSubmit} />
  );

  await act(async () => {
    await true;
    jest.runAllTimers();
  });
  await act(async () => {
    await true;
    jest.runAllTimers();
  });

  await act(async () => {
    root.root.findByType(InstructionSelection).props.onSubmit('emit');
  });

  expect(root.root.findAllByType(InstructionSelection).length).toBe(0);

  act(() => {
    root.root.findByType(InstructionValue).props.onChange({
      foo: 'bar',
    });
  });
  expect(onChange).toHaveBeenCalledWith({
    emit: {
      foo: 'bar',
    },
  });
});
