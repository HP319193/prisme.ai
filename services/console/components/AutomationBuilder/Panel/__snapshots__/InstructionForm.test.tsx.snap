// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="flex flex-1 flex-column"
>
  <div
    className="flex flex-1 align-items-stretch flex-column mb-4"
  >
    <div>
      automations.edit.select
    </div>
    <span
      className="flex p-input-icon-left p-input-icon-right"
    >
      <i
        className="pi pi-search"
      />
      <input
        autoFocus={true}
        className="p-inputtext p-component flex-1"
        onChange={[Function]}
        onInput={[Function]}
        onKeyPress={[Function]}
        placeholder="automations.instruction.search"
        value=""
      />
      <i>
        <button
          onClick={[Function]}
          style={
            Object {
              "background": "none",
              "border": 0,
              "color": "inherit",
            }
          }
        >
          <i
            className="pi pi-times"
          />
        </button>
      </i>
    </span>
  </div>
</div>
`;
